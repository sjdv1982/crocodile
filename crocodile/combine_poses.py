import numpy as np

def combine_poses(poses, format, *, posefilenames = None, indices=None):
    if posefilenames is not None:
        assert len(posefilenames) == len(poses)
    else:
        posefilenames = range(len(poses))
    
    for arr, posefile in zip(poses, posefilenames):
        if arr.dtype not in (np.dtype(float), np.float32):
            raise TypeError("{}: dtype must be float, not {}".format(posefile, arr.dtype))
        if format == "euler":
            dtype = np.dtype([("index", np.uint32),("euler", arr.dtype, (3,))])
            if arr.ndim != 2 or arr.shape[-1] != 3:
                raise TypeError("{}: shape for euler cannot be {}".format(posefile, arr.shape))
        elif format == "quaternion":
            dtype = np.dtype([("index", np.uint32),("quaternion", arr.dtype, (4,))])
            if arr.ndim != 2 or arr.shape[-1] != 4:
                raise TypeError("{}: shape for quaternion cannot be {}".format(posefile, arr.shape))
        elif format == "matrix":
            if arr.ndim != 3 or arr.shape[-2:] not in ((3,3),(4,4)):
                raise TypeError("{}: shape for matrix cannot be {}".format(posefile, arr.shape))
            dtype = np.dtype([("index", np.uint32),("matrix", arr.dtype, arr.shape[-2:])])
    
    totlen = sum([len(arr) for arr in poses])
    result = np.empty(totlen, dtype=dtype)
    pos = 0
    data_prop = format
    if indices is None:
        indices = list(range(len(poses)))
    for narr, arr in enumerate(poses):
        nextpos = pos + len(arr)
        chunk = result[pos:nextpos]
        chunk["index"] = indices[narr]
        chunk[data_prop] = arr
        pos = nextpos
    return result

if __name__ == "__main__":
    import argparse
    import re
    p = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter)
    p.add_argument("format", help="""Format of the input numpy arrays""", 
    choices=["euler", "quaternion", "matrix"])
    p.add_argument("poses", nargs="+", help="""Pose files (numpy format).

The shape must be as follows:
euler: (N, 3)
quaternion: (N, 4)
matrix: (N, 3, 3) or (N, 4, 4)

The dtype must be float or float32"""
    )
    p.add_argument("--indices", required=False, help="""The file index where the pose came from.

If specified, must be a string of numbers separated by spaces or newlines.
By default, indices starting with 0 until (number of files) - 1.""")
    p.add_argument("--output", "-o", required=True, help="""Output file (numpy format).
Output will be a structured numpy array 'arr' where:

    - 'arr["index"]' will contain the file indices  
      (arr[0]["index"] for the first file index)

    - 'arr[data]' will contain the poses 
      (arr[0][data] for the first file pose).
      The 'data' variable is "euler", "quaternion" or "matrix", e.g. arr["euler"]
""") 
    args = p.parse_args()

    poses = []
    for posefile in args.poses:
        arr = np.load(posefile)
        poses.append(arr)
    if args.indices:
        indices = re.split(',|\n| ', args.indices)
        indices = [int(index) for index in indices if len(index)]
    else:
        indices = None
    result = combine_poses(poses, args.format, posefilenames=args.poses, indices=indices)
    np.save(args.output, result)