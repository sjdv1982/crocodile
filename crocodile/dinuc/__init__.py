import numpy as np

dinuc_dtype = np.dtype(
    [
        ("first_resid", np.uint16),
        ("sequence", "S2"),
        ("conformer", np.uint16),
        ("matrix", np.float64, (4, 4)),
        ("rmsd", np.float32),
    ],
    align=True,
)


dinuc_roco_dtype = np.dtype(
    [
        ("first_resid", np.uint16),
        ("sequence", "S2"),
        ("conformer", np.uint16),
        ("rotation_matrix", np.float64, (3, 3)),
        ("offset", np.float64, 3),
        ("rmsd", np.float32),
    ],
    align=True,
)

from .from_ppdb import from_ppdb, ppdb2nucseq
