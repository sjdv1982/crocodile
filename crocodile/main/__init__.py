import numpy as np

GRIDSPACING = np.sqrt(3) / 3

from .main import main, import_config, load_commands, locate_files
