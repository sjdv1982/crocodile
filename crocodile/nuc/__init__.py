import numpy as np
from typing import Tuple
from nefertiti.functions.parse_pdb import atomic_dtype


_basic_mapping_rna = {k: k for k in ("A", "C", "G", "U")}
_basic_mapping_dna = {k: k for k in ("A", "C", "G", "T")}
_basic_mapping_rna.update({"R" + k: k for k in ("A", "C", "G", "U")})
_basic_mapping_dna.update({"D" + k: k for k in ("A", "C", "G", "T")})


def map_resname(
    resname: bytes | str, *, rna: bool = True, extra_mapping: dict | None = None
):
    """Maps a parsed PDB resname to a mononucleotide sequence.
    An unknown mapping gives rise to a KeyError.
    Extra resnames (i.e mutated bases) can be provided via extra_mapping"""

    if isinstance(resname, bytes):
        resname = resname.decode()
    mapping = _basic_mapping_rna if rna else _basic_mapping_dna
    if extra_mapping:
        mapping = mapping.copy().update(extra_mapping)
    try:
        return mapping[resname]
    except KeyError as exc:
        if not rna and resname in _basic_mapping_rna:
            raise KeyError("Mode is DNA (rna=False) but resname is RNA") from None
        elif rna and resname in _basic_mapping_dna:
            raise KeyError("Mode is RNA but resname is DNA") from None
        else:
            raise exc from None


def ppdb2nucseq(
    ppdb: np.ndarray,
    *,
    rna: bool,
    ignore_unknown: bool = False,
    return_index: bool = False,
) -> str | Tuple[str, list]:
    """Converts a nucleic acid parsed PDB to a mono-nucleotide sequence.
    If rna, the parsed PDB is interpreted as RNA, else as DNA
    If ignore_unknown, unknown residue names (i.e. non-canonical bases) are ignored, else raise an exception
    If return_index, return a list of (first, last+1) indices,
      corresponding to the first and last atom indices of each nucleotide
    """
    if ppdb.dtype != atomic_dtype:
        raise TypeError(ppdb.dtype)
    if len(np.unique(ppdb["chain"])) > 1:
        raise ValueError("Multiple chains in parsed PDB")

    indices0 = np.nonzero(np.diff(ppdb["resid"], prepend=-1, append=-1))[0]
    indices = []
    nucs = []
    for n in range(len(indices0) - 1):
        ind1, ind2 = indices0[n], indices0[n + 1]
        try:
            nuc = map_resname(ppdb[ind1]["resname"], rna=rna)
        except KeyError as exc:
            if ignore_unknown:
                continue
            else:
                raise exc from None
        nucs.append(nuc)
        indices.append((ind1, ind2))

    seq = "".join(nucs)
    if return_index:
        return seq, indices
    else:
        return seq
