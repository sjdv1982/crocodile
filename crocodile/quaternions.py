
import numpy as np


def _marsaglia_pairs(n):
    """Generate n pairs x,y uniformly in (-1..1) where xÂ²+yÂ² < 1"""    
    nn = (1 / (np.pi / 4)) * n  # best estimate of how many points we need
    nn += 0.1 * n + 1000  # just to be sure
    while 1:  # just to be really really sure
        pairs = np.random.random_sample(size=(int(nn), 2)) * 2 - 1
        z = pairs[:, 0] ** 2 + pairs[:, 1] ** 2
        pairs_ok = pairs[z < 1]
        if len(pairs_ok) >= n:
            return pairs_ok[:n], z[z<1][:n]

def random_quaternions(n, seed=0):
    #  Generate random quaternions using Marsaglia's method
    np.random.seed(seed)
    xy, z = _marsaglia_pairs(n)
    uv, w = _marsaglia_pairs(n)
    s = np.sqrt((1-z)/w)
    q = np.empty((n, 4))
    q[:, :2] = xy
    q[:, 2:] = s[:, None] * uv
    return q

def quaternions_to_3x3(q):
    """From:
    https://www.euclideanspace.com/maths/geometry/rotations/conversions/quaternionToMatrix/index.htm
    """
    qw, qx, qy, qz = q[:, 0], q[:, 1], q[:, 2], q[:, 3]
    qx2 = qx * qx
    qy2 = qy * qy
    qz2 = qz * qz
    m11 = 1 - 2*qy2 - 2*qz2
    m12 = 2*qx*qy - 2*qz*qw
    m13 = 2*qx*qz + 2*qy*qw
    m21 = 2*qx*qy + 2*qz*qw
    m22 = 1 - 2*qx2 - 2*qz2
    m23 = 2*qy*qz - 2*qx*qw
    m31 = 2*qx*qz - 2*qy*qw
    m32 = 2*qy*qz + 2*qx*qw
    m33 = 1 - 2*qx2 - 2*qy2
    m = np.empty((len(q), 3, 3))
    m[:, 0, 0] = m11
    m[:, 0, 1] = m12
    m[:, 0, 2] = m13
    m[:, 1, 0] = m21
    m[:, 1, 1] = m22
    m[:, 1, 2] = m23
    m[:, 2, 0] = m31
    m[:, 2, 1] = m32
    m[:, 2, 2] = m33
    return m


