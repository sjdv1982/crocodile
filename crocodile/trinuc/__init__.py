import numpy as np

trinuc_dtype = np.dtype(
    [
        ("first_resid", np.uint16),
        ("sequence", "S3"),
        ("conformer", np.uint16),
        ("matrix", np.float64, (4, 4)),
        ("rmsd", np.float32),
    ],
    align=True,
)


trinuc_roco_dtype = np.dtype(
    [
        ("first_resid", np.uint16),
        ("sequence", "S3"),
        ("conformer", np.uint16),
        ("rotation_matrix", np.float64, (3, 3)),
        ("offset", np.float64, 3),
        ("rmsd", np.float32),
    ],
    align=True,
)

_basic_mapping_rna = {k: k for k in ("A", "C", "G", "U")}
_basic_mapping_dna = {k: k for k in ("A", "C", "G", "T")}
_basic_mapping_rna.update({"R" + k: k for k in ("A", "C", "G", "U")})
_basic_mapping_dna.update({"D" + k: k for k in ("A", "C", "G", "T")})


def map_resname(
    resname: bytes | str, *, rna: bool = True, extra_mapping: dict | None = None
):
    """Maps a parsed PDB resname to a mononucleotide sequence.
    An unknown mapping gives rise to a KeyError.
    Extra resnames (i.e mutated bases) can be provided via extra_mapping"""

    if isinstance(resname, bytes):
        resname = resname.decode()
    mapping = _basic_mapping_rna if rna else _basic_mapping_dna
    if extra_mapping:
        mapping = mapping.copy().update(extra_mapping)
    try:
        return mapping[resname]
    except KeyError as exc:
        if not rna and resname in _basic_mapping_rna:
            raise KeyError("Mode is DNA (rna=False) but resname is RNA") from None
        elif rna and resname in _basic_mapping_dna:
            raise KeyError("Mode is RNA but resname is DNA") from None
        else:
            raise exc from None


from .from_ppdb import from_ppdb, ppdb2nucseq
