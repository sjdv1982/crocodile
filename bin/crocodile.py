import sys, os

# Remove current path from imports
currdir = os.path.realpath(os.path.dirname(sys.argv[0]))
sys.path = [d for d in sys.path if os.path.realpath(d) != currdir]

from crocodile.main import main
main()