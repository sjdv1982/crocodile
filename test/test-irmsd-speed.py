import numpy as np
from numpy import sqrt
from numpy.linalg import svd, det
from crocodile.parse_pdb import parse_pdb
receptor = parse_pdb(open("rotate-A.pdb").read())
ligand = parse_pdb(open("rotate-B.pdb").read())
refe_receptor = parse_pdb(open("1AVXA-bound-heavy.pdb").read())
refe_ligand = parse_pdb(open("1AVXB-bound-heavy.pdb").read())

from crocodile.imask import imask
imasks = imask([refe_receptor, refe_ligand], cutoff=10)

def get_coor(pdb):
    coor = np.zeros((len(pdb), 3))
    coor[:, 0] = pdb["x"]
    coor[:, 1] = pdb["y"]
    coor[:, 2] = pdb["z"]
    return coor

def calc_covar(atoms1, atoms2):
    assert atoms1.shape == atoms2.shape
    return atoms1.T.dot(atoms2)

if_r, if_l = get_coor(receptor[imasks[0]]), get_coor(ligand[imasks[1]])
refe_if_r, refe_if_l = get_coor(refe_receptor[imasks[0]]), get_coor(refe_ligand[imasks[1]])

lenr, lenl = len(refe_if_r), len(refe_if_l)
natoms = lenr + lenl    

refe_com = (refe_if_r.sum(axis=0) + refe_if_l.sum(axis=0)) / natoms
refe_if_r -= refe_com
refe_if_l -= refe_com
refe_residual = (refe_if_r*refe_if_r).sum() + (refe_if_l*refe_if_l).sum()

ini_com = (if_r.sum(axis=0) + if_l.sum(axis=0)) / natoms
if_r -= ini_com
if_l -= ini_com

r_sum = if_r.sum(axis=0)
l_sum = if_l.sum(axis=0)
refe_r_sum = refe_if_r.sum(axis=0)
refe_l_sum = refe_if_l.sum(axis=0)

covar_r = calc_covar(if_r, refe_if_r)
covar_l = calc_covar(if_l, refe_if_l)

residual_r = (if_r*if_r).sum()
residual_l = (if_l*if_l).sum()


def calc_irmsd(offset):
    vec1 = -offset * lenl/natoms
    vec2 = offset + vec1

    residual1 = residual_r + (2 * vec1 * r_sum).sum() + (lenr * vec1 * vec1).sum()
    residual2 = residual_l + (2 * vec2 * l_sum).sum() + (lenl * vec2 * vec2).sum()
    residual = residual1 + residual2    

    covar1 = covar_r + vec1[:, None] * refe_r_sum[None, :]
    covar2 = covar_l + vec2[:, None] * refe_l_sum[None, :]
    covar = covar1 + covar2

    v, s, wt = svd(covar)
    if det(v) * det(wt) < 0:
        s[-1] *= -1
        #v[:, -1] *= -1
    ss = (residual + refe_residual) - 2 * s.sum()
    if ss < 0:
        ss = 0
    irmsd = sqrt(ss / natoms)
    return irmsd

def calc_irmsds(offsets):
    vec1 = -offsets * lenl/natoms
    vec2 = offsets + vec1

    residual1 = residual_r + (2 * vec1 * r_sum).sum(axis=1) + (lenr * vec1 * vec1).sum(axis=1)
    residual2 = residual_l + (2 * vec2 * l_sum).sum(axis=1) + (lenl * vec2 * vec2).sum(axis=1)
    residual = residual1 + residual2

    covar1 = covar_r + vec1[:, :, None] * refe_r_sum[None, None, :]
    covar2 = covar_l + vec2[:, :, None] * refe_l_sum[None, None, :]
    covar = covar1 + covar2

    v, s, wt = np.linalg.svd(covar)
    reflect = np.linalg.det(v) * np.linalg.det(wt)
    s[:,-1] *= reflect
    ss = np.maximum( (residual + refe_residual) - 2 * s.sum(axis=1), 0)
    irmsd = np.sqrt(ss / natoms)
    return irmsd

import itertools
offsets = []
for offset0 in itertools.product(range(10), repeat=3):
    offset = np.array(offset0) * 1.2
    offsets.append(offset)
offsets = np.array(offsets)
offsets = np.concatenate([offsets] * 500) # 500 000 calculations

import time
t = time.time()

irmsds = calc_irmsds(offsets)
print("Time elapsed: {:.3f} seconds".format(time.time() - t))
print(irmsds.shape)
for irmsd in irmsds[:10]:
    print(irmsd)