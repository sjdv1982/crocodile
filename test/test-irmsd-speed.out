Time elapsed: 5.075 seconds
(500000,)
1.4940444588341288
1.5378052216671352
1.7288492809754163
2.0278001078463546
2.3961015676518906
2.807830457215913
3.2475381188012924
3.7061406322550807
4.178149445534634
4.660114306256267
